node/network/availability-recovery/src/lib.rs

```diff
@@ -442,59 +419,99 @@ fn reconstructed_data_matches_root(
 	branches.root() == *expected_root
 }
 
-impl Interaction {
-	async fn run(mut self) -> error::Result<()> {
+impl<S: SubsystemSender> Interaction<S> {
+	async fn run(mut self) -> Result<AvailableData, RecoveryError> {
 		loop {
 			// These only fail if we cannot reach the underlying subsystem, which case there is nothing
 			// meaningful we can do.
 			match self.phase {
 				InteractionPhase::RequestFromBackers(ref mut from_backers) => {
-					if from_backers.run(&self.params, &mut self.to_state).await
-						.map_err(error::Error::ClosedToState)?
-					{
-						break Ok(())
-					} else {
-						self.phase = InteractionPhase::RequestChunks(
-							RequestChunksPhase::new(self.params.validators.len() as _)
-						);
+					match from_backers.run(&self.params, &mut self.sender).await {
+						Ok(data) => break Ok(data),
+						Err(RecoveryError::Invalid) => break Err(RecoveryError::Invalid),
+						Err(RecoveryError::Unavailable) => {
+							self.phase = InteractionPhase::RequestChunks(
+								RequestChunksPhase::new(self.params.validators.len() as _)
+							)
+						}
 					}
 				}
 				InteractionPhase::RequestChunks(ref mut from_all) => {
-					break from_all.run(&self.params, &mut self.to_state).await
-						.map_err(error::Error::ClosedToState)
+					break from_all.run(&self.params, &mut self.sender).await;
 				}
 			}
 		}
 	}
 }
 
+/// Accumulate all awaiting sides for some particular `AvailableData`.
+struct InteractionHandle {
+	candidate_hash: CandidateHash,
+	remote: RemoteHandle<Result<AvailableData, RecoveryError>>,
+	awaiting: Vec<oneshot::Sender<Result<AvailableData, RecoveryError>>>,
+}
+
+impl Future for InteractionHandle {
+	type Output = Option<(CandidateHash, Result<AvailableData, RecoveryError>)>;
+
+	fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+		let mut indices_to_remove = Vec::new();
+		for (i, awaiting) in self.awaiting.iter_mut().enumerate().rev() {
+			if let Poll::Ready(()) =  awaiting.poll_canceled(cx) {
+				indices_to_remove.push(i);
+			}
+		}
+
+		// these are reverse order, so remove is fine.
+		for index in indices_to_remove {
+			tracing::debug!(
+				target: LOG_TARGET,
+				candidate_hash = ?self.candidate_hash,
+				"Receiver for available data dropped.",
+			);
+
+			self.awaiting.remove(index);
```

Comment:
that's `O(n^2)` in the worst case, we can do this in `O(n)` by using `swap_remove`
as we do in https://github.com/paritytech/polkadot/blob/726db12a487c08589a1ef2aa32e03b971a150d04/node/network/bridge/src/validator_discovery.rs#L282-L284

as long as we don't care about the ordering

Reply:
The size here will generally be quite small; usually one, in fact. but I will make this change.